File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-5.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        
        // get LinkedList length
        int size = 0;
        for (ListNode* node = head; node; node = node->next) {
            size++;
        }

        // get prev node
        ListNode* prev = head;
        int index = 0;
        int target = size - n - 1;
        while (index < target) {
            prev = prev->next;
            index++;
        }

        // detach
        if (target == -1) {
            return head->next;
        }
        prev->next = prev->next->next;
    
        return head;

    }
};

File: Data Structures & Algorithms/add-two-numbers/submission-5.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode();
        ListNode head = dummy;
        int carry = 0;
        while (l1 != null || l2 != null) {
            int first = l1 != null ? l1.val : 0;
            int second = l2 != null ? l2.val : 0;
            int sum = first + second + carry;
            int value;
            if (sum >= 10) {
                carry = 1;
                value = sum - 10;
            } else {
                carry = 0;
                value = sum;
            }
            ListNode node = new ListNode(value);
            head.next = node;
            head = head.next;
            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
        }
        if (carry != 0) {
            ListNode carryNode = new ListNode(carry);
            head.next = carryNode;
        }
        return dummy.next;
    }
}
